contrib/lib-autopilot.py

```diff
@@ -0,0 +1,289 @@
+        """
+        if k < 3:
+            k = 3
+            
+        self.__logger.info("CENTRAL_NODES: Try to seek {} nodes which are currently central".format(k))        
+        res = [n for n,_ in sorted(nx.betweenness_centrality(self.G).items(),key=itemgetter(1),reverse=True)[:4*k]]
+        self.__logger.info("CENTRAL_NODES: Generated top {} central nodes (according to betweeness centrality)".format(len(res)))
+        
+        self.__logger.info("CENTRAL_NODES: Sample {} items from the candidates as was requested".format(k))
+        tmp = list(res)
+        random.shuffle(tmp)
+        return tmp[0:k]
+
+    def __generate_random_nodes(self,k=3):
+        """
+            generates a random set of nodes.
+            
+            this assumes a uniform distribution of all nodes. 
+        """
+        if k < 3:
+            k = 3
+
+        k = min(k,len(self.G.nodes()))
+        self.__logger.info("RANDOM_NODES: try to generate a set of {} nodes sampled with uniform distribution".format(k))
+        
+        return random.sample(self.G.nodes(),k)
+    
+    def __generate_connicticity_increasing_canidates(self,k=3):
+        """
+            generates a set of nodes that are not well connected
+            for this the start end end nodes of longest paths are retrieved
+        """
+        result = set()
+        
+        if k < 3:
+            k = 3
+        
+        self.__logger.info("IMPROOVE_NETWORK: Try to seek {} nodes which are currently bad connected".format(k))
+        candidates = heapq.nlargest(3*k, self.__generate_all_shortest_paths(), key=itemgetter(0))
+
+        for _, path in candidates:
+            result.add(path[0])
+            result.add(path[-1])
+
+        self.__logger.info("IMPROOVE_NETWORK: Found {} candidates which are currently bad connected".format(len(result)))        
+        if len(result) <= k:
+            return list(result)
+        self.__logger.info("IMPROOVE_NETWORK: Sample {} items from the candidates as was requested".format(k))
+        tmp = list(result)
+        random.shuffle(tmp)
+        return tmp[0:k]
+        
+ 
+    def __generate_all_shortest_paths(self,cutoff = 10):
+        """ 
+        this is basically a wrapper for nx.all_pair_shortest_path to have a more convenient output format
+        
+        generates all shortest paths and yields each item, the format of the output is
+        (len, [src, n1,...,n_{m-2}, dest])
+        """
+        if cutoff < 1:
+            cutoff = 10
+            self.__logger.info("cutoff value must be a positive integer. Set back to default value: 10")
+
+        all_pair_shortest_paths = nx.all_pairs_shortest_path(self.G, cutoff=cutoff)
+        for item in all_pair_shortest_paths:
+            from_node = item[0]
+            paths = item[1]
+            for destination,path in paths.items():
```

Comment:
These 3 lines can be compressed as follows:

```
for src, paths in all_pair_shortest_paths:
```

This just unpacks the item directly into the variables that you'd use below.

Reply:
You actually do that in the line below :wink: